### PR TITLE
tpm2_policyor: add tool

### LINKED ISSUE #863: Implement policyor
Implement a tool to use the policyor command.

### PR DESCRIPTION
**Pull request by idesai**
New tool to compound (OR) policies

Closes #863

### CODE REVIEW

File: lib/tpm2_policy.c [hunk 1]
```diff
@@ -215,3 +231,112 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
     return true;
 }
 
+bool tpm2_policy_populate_digest_list(char *buf, TPML_DIGEST *policy_list,
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: This looks like an internal API routine, this should probably be static.

File: lib/tpm2_policy.c [hunk 2]
```diff
@@ -215,3 +231,112 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+    uint8_t hash_len = tpm2_alg_util_get_hash_size(hash);
+    const char *str_current;
+    int len_current;
+    int len = strlen(str_list);
+    char buf[1024];
+    bool retval = true;
+
+    do {
+        str_current = str_list;
+        str_list = memchr(str_current, ',', len);
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: This code appears to be tokenizing a string on a comma. This should probably use strtok_r. If str_list originates from the argv[] array (think options), it's guaranteed to be NULL terminated.

I think if you strtok_r you can remove the need for the copy into buf below

File: lib/tpm2_policy.c [hunk 3]
```diff
@@ -215,3 +231,112 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+
+bool tpm2_policy_parse_policy_list(const char *str, TPML_DIGEST *policy_list) {
+
+    size_t len = strlen(str);
+    if (str == NULL || len == 0) {
+        return false;
+    }
+
+    //For policy list: Expecting all policy digests of same hash alg type
+    const char *str_left = memchr(str, ':', len);
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: same, strtok_r

File: man/tpm2_policyor.1.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+[common options](common/options.md)
+
+[common tcti options](common/tcti.md)
+
+[supported hash algorithms](common/hash.md)
+
+[algorithm specifiers](common/alg.md)
+
+# EXAMPLES
```
Comment (williamcroberts) [inline man/tpm2_policyor.1.md]: Same comments as tpm2_policyauthorize apply here.

File: lib/tpm2_policy.c [hunk 5]
```diff
@@ -215,3 +231,83 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+bool tpm2_policy_parse_policy_list(char *str, TPML_DIGEST *policy_list) {
+
+    char *str1, *str2, *token, *subtoken;
+    char *saveptr1, *saveptr2;
+    int j;
+    bool retval;
+
+    const char *delimiter1 = ":";
+    const char *delimiter2 = ",";
+    TPMI_ALG_HASH hash = 0;
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: TPM2_ALG_ERROR or something else.... not 0.

File: lib/tpm2_policy.c [hunk 6]
```diff
@@ -215,3 +231,83 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+
+        policy_list->count++;
+
+        return true;
+}
+
+bool tpm2_policy_parse_policy_list(char *str, TPML_DIGEST *policy_list) {
+
+    char *str1, *str2, *token, *subtoken;
+    char *saveptr1, *saveptr2;
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: One per line, we don't do lists.

File: lib/tpm2_policy.c [hunk 7]
```diff
@@ -215,3 +231,83 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+                hash = tpm2_alg_util_from_optarg(subtoken, tpm2_alg_util_flags_hash);
+                if (hash == TPM2_ALG_ERROR) {
+                    return false;
+                }
+            }
+
+            //Multiple valid policy files
+            if (j > 1) {
+                hash_len = tpm2_alg_util_get_hash_size(hash);
+                retval = tpm2_policy_populate_digest_list(subtoken, policy_list, hash_len);
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: Just pass hash here and let the populate routine call tpm2_alg-util_hash_size(). Then you can kill hash_len.

**Comment by idesai** [comment]
@williamcroberts made the changes and rebased it to current master. Thanks.

File: lib/tpm2_policy.c [hunk 8]
```diff
@@ -311,3 +327,90 @@ bool tpm2_policy_get_digest(TSS2_SYS_CONTEXT *sapi_context,
...
+    char *str1;
+    char *str2;
+    char *token;
+    char *subtoken;
+    char *saveptr1;
+    char *saveptr2;
+    const char *delimiter1 = ":";
+    const char *delimiter2 = ",";
+
+    int j;
```
Comment (williamcroberts) [inline lib/tpm2_policy.c]: unsigned

Comment (idesai) [reply lib/tpm2_policy.c]: stroke_r does not like it

**Comment by williamcroberts** [comment]
@idesai can you rebase on this master and force push back?
```sh
git checkout tpm2_policyor
git pull --rebase origin/master
git push -f idsea tpm2_policyor:tpm2_policyor

**Pull request merged**